File: Examples/OSX/AudioKitDemo/AudioKitDemo/Processing/ProcessingViewController.m
//
//  ProcessingViewController.m
//  AudioKitDemo
//
//  Created by Aurelius Prochazka on 2/23/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "ProcessingViewController.h"

#import "AKFoundation.h"
#import "ConvolutionInstrument.h"

@implementation ProcessingViewController
{
    IBOutlet NSSegmentedControl *sourceSegmentedControl;
    IBOutlet NSButton *maintainPitchSwitch;
    IBOutlet AKPropertySlider *speedSlider;
    IBOutlet AKPropertySlider *pitchSlider;
    IBOutlet AKPropertySlider *dishWellSlider;
    IBOutlet AKPropertySlider *dryWetSlider;

    float pitchToMaintain;

    ConvolutionInstrument *convolver;
    AKAudioFilePlayer *audioFilePlayer;

    BOOL isPlaying;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    audioFilePlayer = [[AKAudioFilePlayer alloc] init];
    [AKOrchestra addInstrument:audioFilePlayer];

    convolver = [[ConvolutionInstrument alloc] initWithInput:audioFilePlayer.output];
    [AKOrchestra addInstrument:convolver];

    speedSlider.property = audioFilePlayer.speed;
    pitchSlider.property = audioFilePlayer.scaling;
    dishWellSlider.property = convolver.dishWellBalance;
    dryWetSlider.property   = convolver.dryWetBalance;

    pitchToMaintain = 1.0;
    isPlaying = NO;
}

- (void)viewWillDisappear {
    [self stop:self];
}


- (IBAction)start:(id)sender {
    if (!isPlaying) {
        [convolver start];
        [audioFilePlayer play];
        isPlaying = YES;
    }
}

- (IBAction)stop:(id)sender {
    if (isPlaying) {
        [audioFilePlayer stop];
        [convolver stop];
        isPlaying = NO;
    }
}

- (IBAction)speedChanged:(NSSlider *)sender
{
    if (maintainPitchSwitch.state == NSOnState && fabs(audioFilePlayer.speed.value) > 0.1) {
        audioFilePlayer.scaling.value =  pitchToMaintain / fabs(audioFilePlayer.speed.value);
        [AKTools setSlider:pitchSlider withProperty:audioFilePlayer.scaling];
    }
}

- (IBAction)togglePitchMaintenance:(NSButton *)sender {
    if (sender.state == NSOnState) {
        [pitchSlider setEnabled:NO];
        pitchToMaintain = fabs(audioFilePlayer.speed.value) * audioFilePlayer.scaling.value;
    } else {
        [pitchSlider setEnabled:YES];
    }
}

- (IBAction)fileChanged:(NSSegmentedControl *)sender {
    audioFilePlayer.sampleMix.value = (float) sender.selectedSegment;
}

@end

File: Examples/OSX/Swift/AudioKitDemo/AudioKitDemo/ProcessingViewController.swift
//
//  ProcessingViewController.swift
//  AudioKitDemo
//
//  Created by Nicholas Arner on 3/1/15.
//  Copyright (c) 2015 AudioKit. All rights reserved.
//


class ProcessingViewController: NSViewController {

    @IBOutlet var sourceSegmentedControl: NSSegmentedControl!
    @IBOutlet var maintainPitchSwitch: NSButton!
    @IBOutlet var pitchSlider: NSSlider!

    var isPlaying = false

    var pitchToMaintain: Float = 1.0

    var conv: ConvolutionInstrument!
    let audioFilePlayer = AKAudioFilePlayer()

    override func viewDidLoad() {
        super.viewDidLoad()

        conv = ConvolutionInstrument(input: audioFilePlayer.output)

        AKOrchestra.addInstrument(audioFilePlayer)
        AKOrchestra.addInstrument(conv)
    }

    @IBAction func start(sender:NSButton) {
        if (!isPlaying) {
            conv.play()
            audioFilePlayer.play()
            isPlaying = true
        }
    }

    @IBAction func stop(sender:NSButton) {
        if (isPlaying) {
            conv.stop()
            audioFilePlayer.stop()
            isPlaying = false
        }
    }

    @IBAction func wetnessChanged(sender:NSSlider) {
        AKTools.setProperty(conv.dryWetBalance, withSlider: sender)
    }

    @IBAction func impulseResponseChanged(sender:NSSlider) {
        AKTools.setProperty(conv.dishWellBalance, withSlider: sender)
    }

    @IBAction func speedChanged(sender:NSSlider) {
        AKTools.setProperty(audioFilePlayer.speed, withSlider: sender)
        if (maintainPitchSwitch.state == 1 && fabs(audioFilePlayer.speed.floatValue) > 0.1) {
            audioFilePlayer.scaling.floatValue = pitchToMaintain / fabs(audioFilePlayer.speed.floatValue)
            AKTools.setSlider(pitchSlider, withProperty: audioFilePlayer.scaling)
        }
    }

    @IBAction func pitchChanged(sender:NSSlider) {
        AKTools.setProperty(audioFilePlayer.scaling, withSlider: sender)
    }

    @IBAction func togglePitchMaintenance(sender:NSButton) {
        if sender.state == 1 {
            pitchSlider.enabled = false
            pitchToMaintain = fabs(audioFilePlayer.speed.floatValue) * audioFilePlayer.scaling.floatValue
        } else {
            pitchSlider.enabled = true
        }
    }

    @IBAction func fileChanged(sender:NSSegmentedControl) {
        audioFilePlayer.sampleMix.floatValue = Float(sender.selectedSegment)
    }
}
